subsys/bluetooth/host/adv.c

```diff
@@ -392,6 +391,124 @@ static int hci_set_ad_ext(struct bt_le_ext_adv *adv, uint16_t hci_op,
 	return bt_hci_cmd_send_sync(hci_op, buf, NULL);
 }
 
+static int hci_set_adv_ext_fragmented(struct bt_le_ext_adv *adv, uint16_t hci_op,
+				      const struct bt_ad *ad, size_t ad_len)
+{
+	struct bt_hci_cp_le_set_ext_adv_data *set_data;
+	struct net_buf *buf;
+	int err;
+
+	for (size_t i = 0; i < ad_len; i++) {
+
+		const struct bt_data *data = ad[i].data;
+
+		for (size_t j = 0; j < ad[i].len; j++) {
+			size_t len = data[j].data_len;
+			uint8_t type = data[j].type;
+			size_t offset = 0;
+
+			/* We can't necessarily set one AD field in a single step. */
+			while (offset < data[j].data_len) {
+				buf = bt_hci_cmd_create(hci_op, sizeof(*set_data));
+				if (!buf) {
+					return -ENOBUFS;
+				}
+
+				set_data = net_buf_add(buf, sizeof(*set_data));
+				(void)memset(set_data, 0, sizeof(*set_data));
+
+				set_data->handle = adv->handle;
+				set_data->frag_pref = BT_HCI_LE_EXT_ADV_FRAG_DISABLED;
+
+				/* Determine the operation parameter value. */
+				if ((i == 0) && (j == 0) && (offset == 0)) {
+					set_data->op = BT_HCI_LE_EXT_ADV_OP_FIRST_FRAG;
+				} else if ((i == ad_len - 1) && (j == ad[i].len - 1)) {
+					/* The last AD field may be split into
+					 * one or two commands.
+					 */
+					if (offset != 0) {
+						/* We can always set the data in two operations
+						 * Therefore, we know that this is the last.
+						 */
+						set_data->op = BT_HCI_LE_EXT_ADV_OP_LAST_FRAG;
+					} else if (len + 2 <= BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN) {
+						/* First part fits. */
+						set_data->op = BT_HCI_LE_EXT_ADV_OP_LAST_FRAG;
+					} else {
+						/* The data must be split into two
+						 * commands.
+						 */
+						set_data->op = BT_HCI_LE_EXT_ADV_OP_INTERM_FRAG;
+					}
+				} else {
+					set_data->op = BT_HCI_LE_EXT_ADV_OP_INTERM_FRAG;
+				}
+
+				if (offset == 0) {
+					set_data->len = MIN(len + 2,
+							    BT_HCI_LE_EXT_ADV_FRAG_MAX_LEN);
+				} else {
+					/* No need to take min operation here,
+					 * as we can always fit the second part.
+					 */
+					set_data->len = len - offset;
+				}
+
+				if (offset == 0) {
+					set_data->data[0] = len + 1;
+					set_data->data[1] = type;
+					memcpy(&set_data->data[2], data[j].data, set_data->len);
+					offset += set_data->len - 2;
+				} else {
+					memcpy(&set_data->data[0], &data[j].data[offset],
+					       set_data->len);
+					offset += set_data->len;
+				}
+
+				err = bt_hci_cmd_send_sync(hci_op, buf, NULL);
```

Comment:
This could potentially be a lot of commands. But I guess the tradeoff would be complexity in the code vs effeciency in the HCI commands. Have you checked the processing of this if you had to do this over UART for example?
I'm hoping RPMsg is not too bad, or combined build for that matter.

Reply:
I did not consider performance being a big issue here. Setting long advertising data is typically done once or very seldom.
If performance is an issue, I agree that we could use the non-sync variant of sending commands to the controller, but that makes error handling more difficult.

Reply:
Fair enough. Just curious, but if performance becomes an issue this can be rewritten to be send less HCI commands.
No not caching the error is just trouble, send_sync is preferred.